Pytest 3.0 does not accept empty list in fixture

With pytest 2.x it was very easy to use empty lists in fixture if dynamically run test case did not find anything to test. pytest 3.x seems to accept only None values. 

Also when setting empty list in fixture with pytest 3.x the error message is very ugly 

example code:

```python
@pytest.fixture(scope="module", ids=list(), params=list())
def temp(request):
   return request.param

def test_temp(temp):
     pass
```

---
*(edited by @The-Compiler to add code blocks)*

For reference, this is the behavior with 2.9.2:

```
============================================ test session starts =============================================
platform linux -- Python 3.5.2, pytest-2.9.2, py-1.4.31, pluggy-0.3.1
rootdir: /home/florian, inifile: 
plugins: mock-1.2
collected 1 items 

x.py s

========================================= 1 skipped in 0.00 seconds ==========================================
```

With 3.0.0:

```
============================================ test session starts =============================================
platform linux -- Python 3.5.2, pytest-3.0.0, py-1.4.31, pluggy-0.3.1
rootdir: /home/florian, inifile: 
plugins: hypothesis-2.0.0, catchlog-1.2.1
collected 0 items / 1 errors 

=================================================== ERRORS ===================================================
___________________________________________ ERROR collecting x.py ____________________________________________
.venv/lib/python3.5/site-packages/_pytest/runner.py:163: in __init__
    self.result = func()
.venv/lib/python3.5/site-packages/_pytest/main.py:460: in _memocollect
    return self._memoizedcall('_collected', lambda: list(self.collect()))
.venv/lib/python3.5/site-packages/_pytest/main.py:331: in _memoizedcall
    res = function()
.venv/lib/python3.5/site-packages/_pytest/main.py:460: in <lambda>
    return self._memoizedcall('_collected', lambda: list(self.collect()))
.venv/lib/python3.5/site-packages/_pytest/python.py:404: in collect
    return super(Module, self).collect()
.venv/lib/python3.5/site-packages/_pytest/python.py:318: in collect
    res = self.makeitem(name, obj)
.venv/lib/python3.5/site-packages/_pytest/python.py:330: in makeitem
    collector=self, name=name, obj=obj)
.venv/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py:724: in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
.venv/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py:338: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
.venv/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py:333: in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
.venv/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py:595: in execute
    return _wrapped_call(hook_impl.function(*args), self.execute)
.venv/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py:249: in _wrapped_call
    wrap_controller.send(call_outcome)
.venv/lib/python3.5/site-packages/_pytest/python.py:191: in pytest_pycollect_makeitem
    res = list(collector._genfunctions(name, obj))
.venv/lib/python3.5/site-packages/_pytest/python.py:350: in _genfunctions
    self.ihook.pytest_generate_tests(metafunc=metafunc)
.venv/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py:724: in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
.venv/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py:338: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
.venv/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py:333: in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
.venv/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py:596: in execute
    res = hook_impl.function(*args)
.venv/lib/python3.5/site-packages/_pytest/fixtures.py:1028: in pytest_generate_tests
    ids=fixturedef.ids)
.venv/lib/python3.5/site-packages/_pytest/python.py:846: in parametrize
    ids = idmaker(argnames, argvalues, idfn, ids, self.config)
.venv/lib/python3.5/site-packages/_pytest/python.py:933: in idmaker
    for valindex, valset in enumerate(argvalues)]
.venv/lib/python3.5/site-packages/_pytest/python.py:933: in <listcomp>
    for valindex, valset in enumerate(argvalues)]
.venv/lib/python3.5/site-packages/_pytest/python.py:924: in _idvalset
    if ids is None or ids[idx] is None:
E   IndexError: list index out of range
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
========================================== 1 error in 0.52 seconds ===========================================
```

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -8,12 +8,14 @@
   with normal parameters in the same call (`#1832`_).
   Thanks `@The-Compiler`_ for the report, `@Kingdread`_ and `@nicoddemus`_ for the PR.
 
+* Fix internal error when parametrizing tests or fixtures using an empty ``ids`` argument (`#1849`_).
+  Thanks `@OPpuolitaival`_ for the report and `@nicoddemus`_ for the PR.
+
 * Fix loader error when running ``pytest`` embedded in a zipfile.
   Thanks `@mbachry`_ for the PR.
 
 *
 
-*
 
 *
 
@@ -21,9 +23,11 @@
 
 .. _@Kingdread: https://github.com/Kingdread
 .. _@mbachry: https://github.com/mbachry
+.. _@OPpuolitaival: https://github.com/OPpuolitaival
 
 .. _#1822: https://github.com/pytest-dev/pytest/issues/1822
 .. _#1832: https://github.com/pytest-dev/pytest/issues/1832
+.. _#1849: https://github.com/pytest-dev/pytest/issues/1849
 
 
 3.0.0

--- a/_pytest/python.py
+++ b/_pytest/python.py
@@ -936,7 +936,7 @@ def _idval(val, argname, idx, idfn, config=None):
     return str(argname)+str(idx)
 
 def _idvalset(idx, valset, argnames, idfn, ids, config=None):
-    if ids is None or ids[idx] is None:
+    if ids is None or (idx >= len(ids) or ids[idx] is None):
         this_id = [_idval(val, argname, idx, idfn, config)
                    for val, argname in zip(valset, argnames)]
         return "-".join(this_id)

--- a/testing/python/metafunc.py
+++ b/testing/python/metafunc.py
@@ -889,6 +889,33 @@ class TestMetafuncFunctional:
             "*test_function*advanced*FAILED",
         ])
 
+    def test_fixture_parametrized_empty_ids(self, testdir):
+        """Fixtures parametrized with empty ids cause an internal error (#1849)."""
+        testdir.makepyfile('''
+            import pytest
+
+            @pytest.fixture(scope="module", ids=[], params=[])
+            def temp(request):
+               return request.param
+
+            def test_temp(temp):
+                 pass
+        ''')
+        result = testdir.runpytest()
+        result.stdout.fnmatch_lines(['* 1 skipped *'])
+
+    def test_parametrized_empty_ids(self, testdir):
+        """Tests parametrized with empty ids cause an internal error (#1849)."""
+        testdir.makepyfile('''
+            import pytest
+
+            @pytest.mark.parametrize('temp', [], ids=list())
+            def test_temp(temp):
+                 pass
+        ''')
+        result = testdir.runpytest()
+        result.stdout.fnmatch_lines(['* 1 skipped *'])
+
     def test_parametrize_with_identical_ids_get_unique_names(self, testdir):
         testdir.makepyfile("""
             import pytest

